### PR TITLE
Support for new K8s modules

### PR DESCRIPTION
**Pull request by chouseknecht**
##### SUMMARY
Adds the following to support new K8s modules

- Utility module to support the new K8s modules. 
- Kubernetes guide

Individual modules are currently generated from [openshift-restclient-python](https://github.com/openshift/openshift-restclient-python). They are available for pre-release testing via the [ansible.kubernetes-modules](https://galaxy.ansible.com/ansible/kubernetes-modules) role.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
lib/ansible/module_utils/k8s_common.py
docs/docsite/rst/guide_kubernetes.py

##### ANSIBLE VERSION
```
ansible 2.4.0 (k8scommon 04446b5a40) last updated 2017/03/23 05:03:39 (GMT -400)
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.12 (default, Oct 11 2016, 05:24:00) [GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.38)]
```

### CODE REVIEW

File: lib/ansible/module_utils/k8s_common.py [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+        self.argspec_cache = spec
+        return self.argspec_cache
+
+    def execute_module(self):
+        """
+        Performs basic CRUD operations on the model object. Ends by calling
+        AnsibleModule.fail_json(), if an error is encountered, otherwise
+        AnsibleModule.exit_json() with a dict containing:
+          changed: boolean
+          api_version: the API version passsed in
```
Comment (ryansb) [inline lib/ansible/module_utils/k8s_common.py]: s/passs/pass/

**Review by ryansb — COMMENTED** [review]
Looks good, but let me see if I have the deps straight: this requires the OpenShift library for the AnsibleModuleHelper (and other things), then this module utils change, and the actual modules are in the role on Galaxy for now. Is that right?

**Comment by chouseknecht** [comment]
Yes, the individual K8s modules are only available via the role for now, until they get more testing, and stablilize. They all rely on this module to be present in lib/ansible/module_utils, and this relies on the openshift module, which provides AnsibleModuleHelper.

role with K8s modules -> k8s_common.py -> openshift/openshift-restclient-python

**Comment by chouseknecht** [comment]
Working on removing `import string_utils`. That already gets installed by the `openshift`, module.

**Comment by chouseknecht** [comment]
!needs_revision

File: lib/ansible/module_utils/k8s_common.py [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+import copy
+import json
+import os
+import yaml
+
+from ansible.module_utils.basic import AnsibleModule
+from openshift.helper.ansible import AnsibleModuleHelper, OpenShiftException, ARG_ATTRIBUTES_BLACKLIST
```
Comment (bcoca) [inline lib/ansible/module_utils/k8s_common.py]: this is not in core and will result  in exceptions if not present, use same pattern as we do with boto

**Review by bcoca — CHANGES_REQUESTED** [review]

File: lib/ansible/module_utils/k8s_common.py [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+import copy
+import json
+import os
+import yaml
```
Comment (bcoca) [inline lib/ansible/module_utils/k8s_common.py]: yaml is also not core nor requried on targets, should also be 'exception wrapped'

**Review by bcoca — CHANGES_REQUESTED** [review]

File: lib/ansible/module_utils/k8s_common.py [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+    @property
+    def argspec(self):
+        """
+        Build the module argument spec from the helper.argspec, removing any extra attributes not needed by
+        Ansible.
+
+        :return: dict: a valid Ansible argument spec
+        """
+        if self.argspec_cache:
+            return self.argspec_cache
```
Comment (bcoca) [inline lib/ansible/module_utils/k8s_common.py]: `if not self.argspec_cache:` allows you to have a single return, not blocker, just seems cleaner

File: lib/ansible/module_utils/k8s_common.py [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+                return_attributes[self.kind] = k8s_obj.to_dict()
+                self.exit_json(**return_attributes)
+            elif self.helper.has_method('create'):
+                # For a rollback, execute a POST, and exit
+                k8s_obj = self.__create(namespace)
+                return_attributes[self.kind] = k8s_obj.to_dict()
+                return_attributes['changed'] = True
+                self.exit_json(**return_attributes)
+            else:
+                self.fail_json(msg="Missing state parameter. Expected one of: present, absent, replaced.")
```
Comment (bcoca) [inline lib/ansible/module_utils/k8s_common.py]: replaced is not a good state, it should be part of present (would be 'changed' return status)

**Review by bcoca — CHANGES_REQUESTED** [review]

**Comment by bcoca** [comment]
hard to review as a lot of the code is externalized, this also makes the error handling very generic and seemingly not user friendly, but it depends mainly on what the other code is throwing as exceptions.

i.e 'Failed to retrieve requested object' doesn't tell the user if the problem was his input, connection issues, etc nor the exact operational part that failed, just a very generic retrieval failure.

File: lib/ansible/module_utils/k8s_common.py [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+                param_name = self.helper.attribute_to_snake(key)
+            if param_name in self.helper.argspec and value is not None:
+                parameters[param_name] = value
+            elif isinstance(value, dict):
+                continue_path = copy.copy(path) if path else []
+                continue_path.append(self.helper.attribute_to_snake(key))
+                self._add_parameter(value, continue_path, parameters)
+            else:
+                self.fail_json(
+                    msg="Error parsing resource definition. Encountered {}, which does not map to a "
```
Comment (bcoca) [inline lib/ansible/module_utils/k8s_common.py]: does this indicate that the module itself is broken? it should be indicated then that the user cannot fix the problem and that  a bug should be opened.

File: lib/ansible/module_utils/k8s_common.py [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+        """ Load the requested src path """
+        result = None
+        path = os.path.normpath(src)
+        self.helper.log("Reading definition from {}".format(path))
+        if not os.path.exists(path):
+            self.fail_json(msg="Error accessing {}. Does the file exist?".format(path))
+        try:
+            result = yaml.safe_load(open(path, 'r'))
+        except yaml.YAMLError as exc:
+            self.fail_json(msg="Invalid YAML: {}".format(exc))
```
Comment (bcoca) [inline lib/ansible/module_utils/k8s_common.py]: does exc include file information?

**Pull request merged**